dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,177 @@ public Pattern getParameterPattern() {
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDBMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDBMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs) {
+            return Direction.INPUT;
+        }
+    }
+
+    private class DatabaseProcedure extends DatabaseObject {
+
+        private DatabaseProcedure(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDBMetaData().getProcedures(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDBMetaData().getProcedureColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            switch (rs.getInt("COLUMN_TYPE")) {
+                case DatabaseMetaData.procedureColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+                default:
+                    throw new SQLException("Unpexpected procedure parameter type for parameter " +
+                                           rs.getString("COLUMN_NAME"));
+            }
+        }
+    }
+
+    private class DatabaseFunction extends DatabaseObject {
+
+        private DatabaseFunction(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDBMetaData().getFunctions(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDBMetaData().getFunctionColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            int paramType = rs.getInt("COLUMN_TYPE");
+            switch (paramType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+                default:
+                    throw new SQLException("Unpexpected function parameter type for parameter " +
+                                  rs.getString("COLUMN_NAME"));
```

Comment:
I'm thinking of using something like guava's `ImmutableMap.of()` or `ImmutableMap.builder` to initialise a `static Map` from a list of literals. Otherwise with `HashMap` it has to be done in a method (perhaps the `DatabaseProcedure/Function` constructor).

Then we have a method in the new `DatabaseExecutable` (or `DatabaseRoutine`?) that gets the `Direction` value from the JDBC parameter type `int`. If it's not in the map then throw the exception.

It still sounds like I have something like 
```
    if (getParamDirection(jdbcIntValue) == null) {
        throw new SQLException("Invalid parameter type");
    }
    return (getParamDirection(jdbcIntValue);
````